torch_xla/csrc/runtime/computation_client.h

```diff
@@ -310,9 +310,8 @@ class ComputationClient {
   // The result[i], a vector itself, will be the result of the computation fed
   // with arguments[i]. If options.explode_tuple is true, the output tuples will
   // be decomposed into their single elements.
-  virtual std::vector<std::vector<DataPtr>> ExecuteReplicated(
-      const Computation& computation,
-      const std::vector<std::vector<DataPtr>>& arguments,
+  virtual std::vector<DataPtr> ExecuteReplicated(
```

Comment:
Actually, I am very reluctant to change the return type here -- device x arguments (vector of vector) is more general as the original ExecuteReplicated was meant to be implemented for non-SPMD as well. Also, I see that you are removing the input handler and the output handler... those abstractions were introduced to avoid any assumption about the computation client's return type, other than it's a device Data. The actual sharidng related processing happens in the sharding util, only after realizing that it's a sharded data.

Any reason you need to refactor the SPMD related code path, under pjrt_computation_client and xla_sharding_util? I assume this is for IFRT migration, and you want to skip having a separate ifrt client?

Reply:
I don't think there's a reason to keep the "vector of vector" structure here. We may have supported some way of running Replicated execution without SPMD before with XRT, but it doesn't really exist in PJRT or IFRT. You either execute on one device (PJRT's ExecuteSharded) or all local devices (IFRT's Execute). There was actually no usage of ExecuteReplicated at all within PT/XLA before PJRT and SPMD. IIRC we only kept this interface for this long to keep compatibility with XRT (even though we hadn't actually used that code path in XRT for years, if ever). We explicitly only call `ExecuteReplicated` when `async->cached_computation->is_sharded`.

It doesn't look like `InputHandler` would have handled unsharded data (what would `GetDataShard` do to unsharded data?), nor does it look like `OutputHandler` would have handled sharded data (what would `WrapDataShards` do to sharded data?). I think it makes much more sense to just construct the sharded data within the `ComputationClient`, applying the `xla::OpSharding`s from the actual `PjRtExecutable`. It's a relatively minor refactor in PJRT, and it makes IFRT much easier to implement.

Reply:
Yes, InputHandler and OuptutHandler works with both unsharded and sharded, and it's a no-op for unsharded data. We were able to switch around the data handling logic without touchting the data representation, and also the vice versa. Having that abstraction layer was the point I was trying to make, not the functional perspective ---- it works either way.

So for me, the question really comes down to if we are creating a separate IFRT computation client, and then later make it the main/default client for all use cases.

Reply:
Also ExecuteReplicate really calls Execute from the runtime, which has nested buffer... https://source.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/pjrt/pjrt_client.h;l=1373?q=Execute

Now for us, we have a sharded data (or ifrt array) that wraps a vector of buffer, so we can make it either way -- but, assuming that the computation data could just be a wrapper of the device buffer -- the former is not strange at all?

Reply:
Having said that, if you feel that this would make the IFRT adoption easier, then let's make this refactor functional and land.

Reply:
The fundamental difference between PJRT and IFRT in this case is that it's trivial to "unshard" data in PJRT, since the actual runtime construct (`PjRtBuffer`) doesn't know anything about sharding. All sharding information is effectively in the framework (`PjRtShardedData`) and in the compiled `PjRtExecutable`.

In IFRT, the runtime data construct (`ifrt::Array`) carries its own sharding information. Disassembling it actually creates new `ifrt::Array`s with new shardings, then we'd have to reassemble the array with the original sharding before passing it into the `ifrt::Executable`. The executable will then return a sharded array, which we'll again have to disassemble, then reassemble in `OutputHandler` with the sharding that matches what the executable returned. This creates a lot of space for errors to hide, particularly around the ownership of the underlying shards of data, which will be aliased by multiple `ifrt::Array`s (compared to PJRT, where a single buffer of device data is always owned by one `PjRtBuffer`).

To avoid this, we either need to 1) implement the hack in my IFRT prototype PR where I skip `InputHandler` and `OutputHandler` when using IFRT or 2) abstract more implementation details under `ComputationClient`, since the data representations need to be handled differently.

Since `DataPtr` already represents sharded data, it's actually very easy for us to approximate IFRT's behavior with PJRT. `PjRtShardedData` fills the same role as `ifrt::Array`: references to the local device data plus the sharding of that data. It would be much harder for us to approximate PjRt's behavior from IFRT.

This PR is definitely inspired by IFRT, but I also think this is the right way to evolve this abstraction post-XRT. It doesn't really make the runtime client any thicker, and overall it pushes functionality down in the stack (ie out of PyTorch/XLA) now that we have more capable runtimes.

Reply:
I hoped that the SPMD sharding logics/functions to be runtime agnostic, working with the PjRtShardedData & DataPtr abstraction. Question, ifrt::Array tracks references or the actual copies? If it's the former like the PjRtBuffer, then assemble, disassemble shouldn't be a problem? One thing for sure, we don't want to materialize those data to CPU in any time. The input handler & output handler in SPMD util just reorganizes references. 

Now come to think of it, yea -- adopting ifrt::Array is the right way, and it doesn't seem to be a simple drop-in replacement. Let's make it work, and at the same time -- I wonder if we could see to make leave `InputHandler` and `OutputHandler` intact so it can work with both PjRt and IfRt? 

I am also going to expand the InputHandler to take care of resharding the inputs after the auto-sharding pass. At the high-level, `InputHandler` is supposed to prepare the inputs in PyTorch/XLA before the execution, `OutputHandler` post-processes the outputs of the execution.

Reply:
Each operation in IFRT that produces a new Array (copying, disassembling, assembling, resharding, etc) allows you to [copy, donate, or reuse the input](https://github.com/openxla/xla/blob/main/xla/python/ifrt/array.h#L39-L54). I don't believe reusing the input actually means the new buffer shares ownership. This is not relevant in PJRT, since it's functionally impossible (by design) to alias a PJRT buffer. One piece of device data is owned by corresponds to one PJRT buffer, and we share ownership with a shared_ptr to that buffer. That's no longer true in IFRT: particularly when we're resharding, we may end up with multiple arrays that correspond to the same device data, and one array owns that data.

The IFRT client will also need to handle one more unique case: for single-controller, Arrays will actually represent both the addressable _and_ non-addressable shards of a global tensor. The representations of sharded data are different enough that the `ComputationClient` implementations will have to treat the sharded data containers differently.

The current function of `InputHandler` (ie converting a vector of objects that each contain a vector of shards, into a vector of vectors of shards) should move down into `PjRtComputationClient`, since that is specific to the API of `PjRtClient::Execute`. Likewise, the inverse operation in `OutputHandler` belongs in `PjRtComputationClient` for the same reason. `OutputHandler` does one more thing: assign shardings to the outputs.

Intuitively, I would say it makes the most sense to assign the global shapes and shardings to each sharded data based on the executable that produced them. I'm pretty sure the `ShardingSpec`s in `OutputHandler` originate from the compiled `Executable`'s `GetHloModules` anyway, and PJRT added [APIs for specifically this purpose](https://github.com/openxla/xla/blob/main/xla/pjrt/pjrt_executable.h#L325-L326) since we started SPMD. But these calls are [far too slow in practice](https://github.com/pytorch/xla/pull/5737#issuecomment-1787696074).

When I went to add back the sharding assignment in `OutputHandler`, I noticed that we don't even actually use the shardings there! We assign all of the outputs to placeholders:

https://github.com/pytorch/xla/blob/b20a082a7c616f9f5d08fa457d7ee4db0a03b52f/torch_xla/csrc/xla_graph_executor.cpp#L749-L756

https://github.com/pytorch/xla/blob/b20a082a7c616f9f5d08fa457d7ee4db0a03b52f/torch_xla/csrc/xla_graph_executor.cpp#L1022-L1025

`Assign` completely ignores the sharding and shape on the right side:

https://github.com/pytorch/xla/blob/b20a082a7c616f9f5d08fa457d7ee4db0a03b52f/torch_xla/csrc/runtime/pjrt_computation_client.h#L192-L198

In commit [5c45ca6](https://github.com/pytorch/xla/pull/5737/commits/5c45ca65544b34e4b72cfcd53f35369d34664473), I just gave all of the outputs a placeholder "manual" sharding, since the placeholders have the real sharding anyway.

Reply:
That makes sense, since we should extract the output (propagated, non-propagated) after the compilation, and they are available when we prepare the placeholders. One thing to note (synced offline) is that the UNKNOWN sharidng type carry a special meaning for the upcoming auto-sharding change. For this PR, we can just focus on avoiding any regressions.

Reply:
As discussed offline, I gave the outputs here "real" shardings based on the compiled executable, since `UNKNOWN` will have a real semantic meaning soon. The extra call to `GetOutputShardings` works here since I only call it once per compilation, rather than once per execution.